In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# from functions import *

In [2]:
# Loading datas

customers = pd.read_csv('olist_customers_dataset.csv', sep=',')
geolocation = pd.read_csv('olist_geolocation_dataset.csv', sep=',')
order_items = pd.read_csv('olist_order_items_dataset.csv', sep=',')
order_payments = pd.read_csv('olist_order_payments_dataset.csv', sep=',')
order_reviews = pd.read_csv('olist_order_reviews_dataset.csv', sep=',')
orders = pd.read_csv('olist_orders_dataset.csv', sep=',')
products = pd.read_csv('olist_products_dataset.csv', sep=',')
sellers = pd.read_csv('olist_sellers_dataset.csv', sep=',')
product_category_name_translation = pd.read_csv('product_category_name_translation.csv', sep=',')

In [3]:
customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [4]:
geolocation.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [5]:
order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [6]:
order_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [7]:
order_reviews.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [8]:
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [9]:
products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [10]:
sellers.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

In [11]:
product_category_name_translation.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

# 1. Merging datasets

We will try to extract all relevant information on customers on one unique dataset

## I - Customers and orders

In [34]:
customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [13]:
customers.shape

(99441, 5)

In [100]:
customers.isna().sum().sum()

0

In [14]:
# Looking duplicates
customers[customers.duplicated(subset='customer_unique_id')]

customer_id                customer_unique_id  \
679    c57b4b6f3719475543b721e720a526ad  b6c083700ca8c135ba9f0f132930d4e8   
1463   9f6f3da49e2d46e3a7529f5e3c25ecce  a40096fc0a3862e9e12bc55b5f8e6ab2   
1607   299f7b5125c8fbe1761a1b320c34fc7d  b8b3c435a58aebd788a477bed8342910   
2811   226d59f9f4b98e228b689eea45d03a6d  66980c3775537536f77b434d74e520f5   
3198   b1253701171dfb298f52a221f824e45b  788313d21c3507fe890921f6e17aa679   
...                                 ...                               ...   
99324  5b46a0d983eec8c97363bea78d4a69dd  8bab3162259edfaadd1ea2e1fe7f58dc   
99327  c1affa46f9f3b514555259049a0307b9  12ab9334b1240d6d037f2b0102a49571   
99336  ebf46ff530343a129926adc1f831dea4  0ee57f62666561b72f2ceacad0230cbf   
99353  282fbce48e4d2077aad602dd125c9225  0ceb502fc33a2ad327b08288c5310e2e   
99406  d9110683c7a282144e9fc97660026a28  5cbfdb85ec130898108b32c50d619c39   

       customer_zip_code_prefix         customer_city customer_state  
679                        8686                suzano             SP  
1463                      20561        rio de janeiro             RJ  
1607                      95585         arroio do sal             RS  
2811                      30290        belo horizonte             MG  
3198                      11070                santos             SP  
...                         ...                   ...            ...  
99324                     31565        belo horizonte             MG  
99327                     38050               uberaba             MG  
99336                      9530    sao caetano do sul             SP  
99353                     29134                 viana             ES  
99406                     74980  aparecida de goiania             GO  

[3345 rows x 5 columns]

We have 3 345 customers which has done 2 or more orders on Olist

In [35]:
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [18]:
orders.shape

(99441, 8)

In [102]:
orders.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [30]:
# We check if we have identical customer_id in df customers and df order

id_1 = list(customers['customer_id'].sort_values())
id_2 = list(orders['customer_id'].sort_values())

id_1 == id_2

True

In [136]:
data = pd.merge(customers, orders, on=['customer_id'], how='outer')
data

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       customer_zip_code_prefix          customer_city customer_state  \
0                         14409                 franca             SP   
1                          9790  sao bernardo do campo             SP   
2                          1151              sao paulo             SP   
3                          8775        mogi das cruzes             SP   
4                         13056               campinas             SP   
...                         ...                    ...            ...   
99436                      3937              sao paulo             SP   
99437                      6764        taboao da serra             SP   
99438                     60115              fortaleza             CE   
99439                     92120                 canoas             RS   
99440                      6703                  cotia             SP   

                               order_id order_status order_purchase_timestamp  \
0      00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1      29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2      b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3      951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4      6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   
...                                 ...          ...                      ...   
99436  6760e20addcf0121e9d58f2f1ff14298    delivered      2018-04-07 15:48:17   
99437  9ec0c8947d973db4f4e8dcf1fbfa8f1b    delivered      2018-04-04 08:20:22   
99438  fed4434add09a6f332ea398efd656a5c    delivered      2018-04-08 20:11:50   
99439  e31ec91cea1ecf97797787471f98a8c2    delivered      2017-11-03 21:08:33   
99440  28db69209a75e59f20ccbb5c36a20b90    delivered      2017-12-19 14:27:23   

         order_approved_at order_delivered_carrier_date  \
0      2017-05-16 15:22:12          2017-05-23 10:47:57   
1      2018-01-12 20:58:32          2018-01-15 17:14:59   
2      2018-05-20 16:19:10          2018-06-11 14:31:00   
3      2018-03-13 17:29:19          2018-03-27 23:22:42   
4      2018-07-29 10:10:09          2018-07-30 15:16:00   
...                    ...                          ...   
99436  2018-04-07 16:08:45          2018-04-11 02:08:36   
99437  2018-04-04 08:35:12          2018-04-05 18:42:35   
99438  2018-04-08 20:30:03          2018-04-09 17:52:17   
99439  2017-11-03 21:31:20          2017-11-06 18:24:41   
99440  2017-12-19 18:50:39          2017-12-21 19:17:21   

      order_delivered_customer_date order_estimated_delivery_date  
0               2017-05-25 10:35:35           2017-06-05 00:00:00  
1               2018-01-29 12:41:19           2018-02-06 00:00:00  
2               2018-06-14 17:58:51           2018-06-13 00:00:00  
3               2018-03-28 16:04:25           2018-04-10 00:00:00  
4               2018-08-09 20:55:48           2018-08-15 00:00:00  
...                             ...                           ...  
99436           2018-04-13 20:06:37           2018-04-25 00:00:00  


## II - Data and order_items

In [82]:
order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [92]:
order_items.shape

(112650, 7)

In [103]:
order_items.isna().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [89]:
# Looking duplicates
order_items[order_items.duplicated(subset='order_id')]

order_id  order_item_id  \
14      0008288aa423d2a3f00fcb17cd7d8719              2   
33      00143d0f86d6fbd9f9b38ab440ac16f5              2   
34      00143d0f86d6fbd9f9b38ab440ac16f5              3   
43      001ab0a7578dd66cd4b0a71f5b6e1e41              2   
44      001ab0a7578dd66cd4b0a71f5b6e1e41              3   
...                                  ...            ...   
112617  ffecd5a79a0084f6a592288c67e3c298              3   
112635  fff8287bbae429a99bb7e8c21d151c41              2   
112641  fffb9224b6fc7c43ebb0904318b10b5f              2   
112642  fffb9224b6fc7c43ebb0904318b10b5f              3   
112643  fffb9224b6fc7c43ebb0904318b10b5f              4   

                              product_id                         seller_id  \
14      368c6c730842d78016ad823897a372db  1f50f920176fa81dab994f9023523100   
33      e95ee6822b66ac6058e2e4aff656071a  a17f621c590ea0fab3d5d883e1630ec6   
34      e95ee6822b66ac6058e2e4aff656071a  a17f621c590ea0fab3d5d883e1630ec6   
43      0b0172eb0fd18479d29c3bc122c058c2  5656537e588803a555b8eb41f07a944b   
44      0b0172eb0fd18479d29c3bc122c058c2  5656537e588803a555b8eb41f07a944b   
...                                  ...                               ...   
112617  50fd2b788dc166edd20512370dac54df  8b321bb669392f5163d04c59e235e066   
112635  bee2e070c39f3dd2f6883a17a5f0da45  4e922959ae960d389249c378d1c939f5   
112641  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   
112642  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   
112643  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   

        shipping_limit_date   price  freight_value  
14      2018-02-21 02:55:52   49.90          13.37  
33      2017-10-20 16:07:52   21.33          15.10  
34      2017-10-20 16:07:52   21.33          15.10  
43      2018-01-04 02:33:42   24.89          17.63  
44      2018-01-04 02:33:42   24.89          17.63  
...                     ...     ...            ...  
112617  2018-03-05 20:15:27   21.90          15.79  
112635  2018-03-27 12:29:22  180.00          48.14  
112641  2017-11-03 02:55:58   55.00          34.19  
112642  2017-11-03 02:55:58   55.00          34.19  
112643  2017-11-03 02:55:58   55.00          34.19  

[13984 rows x 7 columns]

In [93]:
# One order can have several items, let's check it
order_items.loc[order_items['order_item_id'] > 1]

order_id  order_item_id  \
14      0008288aa423d2a3f00fcb17cd7d8719              2   
33      00143d0f86d6fbd9f9b38ab440ac16f5              2   
34      00143d0f86d6fbd9f9b38ab440ac16f5              3   
43      001ab0a7578dd66cd4b0a71f5b6e1e41              2   
44      001ab0a7578dd66cd4b0a71f5b6e1e41              3   
...                                  ...            ...   
112617  ffecd5a79a0084f6a592288c67e3c298              3   
112635  fff8287bbae429a99bb7e8c21d151c41              2   
112641  fffb9224b6fc7c43ebb0904318b10b5f              2   
112642  fffb9224b6fc7c43ebb0904318b10b5f              3   
112643  fffb9224b6fc7c43ebb0904318b10b5f              4   

                              product_id                         seller_id  \
14      368c6c730842d78016ad823897a372db  1f50f920176fa81dab994f9023523100   
33      e95ee6822b66ac6058e2e4aff656071a  a17f621c590ea0fab3d5d883e1630ec6   
34      e95ee6822b66ac6058e2e4aff656071a  a17f621c590ea0fab3d5d883e1630ec6   
43      0b0172eb0fd18479d29c3bc122c058c2  5656537e588803a555b8eb41f07a944b   
44      0b0172eb0fd18479d29c3bc122c058c2  5656537e588803a555b8eb41f07a944b   
...                                  ...                               ...   
112617  50fd2b788dc166edd20512370dac54df  8b321bb669392f5163d04c59e235e066   
112635  bee2e070c39f3dd2f6883a17a5f0da45  4e922959ae960d389249c378d1c939f5   
112641  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   
112642  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   
112643  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   

        shipping_limit_date   price  freight_value  
14      2018-02-21 02:55:52   49.90          13.37  
33      2017-10-20 16:07:52   21.33          15.10  
34      2017-10-20 16:07:52   21.33          15.10  
43      2018-01-04 02:33:42   24.89          17.63  
44      2018-01-04 02:33:42   24.89          17.63  
...                     ...     ...            ...  
112617  2018-03-05 20:15:27   21.90          15.79  
112635  2018-03-27 12:29:22  180.00          48.14  
112641  2017-11-03 02:55:58   55.00          34.19  
112642  2017-11-03 02:55:58   55.00          34.19  
112643  2017-11-03 02:55:58   55.00          34.19  

[13984 rows x 7 columns]

In [108]:
# We should have the same number of orders than products after eliminating rows concerning order_item_id > 1
data.shape[0] - order_items.loc[order_items['order_item_id'] == 1].shape[0]

775

In [120]:
# Let's try to see wich orders doesn't have a product associated

# Creating to lists with unique order_id of df order_items and data
items_unique = list(order_items.loc[order_items['order_item_id'] == 1]['order_id'])
total_orders = list(data['order_id'])

diff = set(total_orders) - set(items_unique) # Creation of list with order_id in data and not in order_items
len(diff) # Checking it corresponds to the deference above

775

In [121]:
data[data['order_id'].isin(diff)]

customer_id                customer_unique_id  \
46     f34a6e874087ec1f0e3dab9fdf659c5d  233896de79986082f1f479f1f85281cb   
288    5bfe800011656c0afb81db64519982db  0071f46a072a9ae25bbe4438b15efe9c   
373    b08064e24083fee8fbe8797902b07ecd  035f60af6e7d7f78470e9443be08d339   
383    536f46cc0f2f2b1e40d056f7998f0254  340152332a04149987a705602615f0d0   
556    8118922685d2e2c0205f060be4f2579c  d0e87d00021530383c16452a39a393ba   
...                                 ...                               ...   
98267  7353b0fb8e8d9675e3a704c60ca44ebe  21c933c8dd97d088e64c50988c90ccf5   
98319  5effe9491d0f5a1c12ee4a384ec09bdf  e9429d622d9e764ca60ba01bc252b0a8   
98551  75c6e03e7206bd1dd59f427c9f75142a  ff36be26206fffe1eb37afd54c70e18b   
98642  1761db2f0d235bbfa1c281093198af0c  21c2c79d24e2d8e3496aabb1f65d85c6   
99105  882b44646ab3f186597b1622a9443ff9  639b4096a4e5c96549a9bb495ca4871e   

       customer_zip_code_prefix   customer_city customer_state  \
46                        38300       ituiutaba             MG   
288                       16011       aracatuba             SP   
373                       28495         aperibe             RJ   
383                       87030         maringa             PR   
556                       45810    porto seguro             BA   
...                         ...             ...            ...   
98267                      5017       sao paulo             SP   
98319                      8690          suzano             SP   
98551                     13057        campinas             SP   
98642                     20270  rio de janeiro             RJ   
99105                     13972         itapira             SP   

                               order_id order_status order_purchase_timestamp  \
46     6e98de3a85c84ead6689189b825d35b5     canceled      2018-03-15 10:07:02   
288    df8c077268f7f3baaac0892eb3143642  unavailable      2017-02-01 00:04:17   
373    c609f82bcf7a90292a5940205ebd7e93  unavailable      2018-05-13 16:45:55   
383    cb4a79c1e6c9ae44302861e7602cc449  unavailable      2017-12-06 15:46:07   
556    af264f3527e94e431f0dcd56cd6b406d  unavailable      2017-07-12 14:49:11   
...                                 ...          ...                      ...   
98267  2f634e2cebf8c0283e7ef0989f77d217  unavailable      2017-09-27 20:55:33   
98319  2b0edc4c59d83dcef85466718c36a317  unavailable      2017-09-29 19:17:10   
98551  81b7c7bbc8ec003eeb67d87441a6a148     canceled      2018-08-20 09:03:23   
98642  8b08f0e729f58529ed03e763270f78d5  unavailable      2017-08-24 21:02:47   
99105  9b932dca249f9971b86dde2f2d7ad412  unavailable      2017-11-20 17:08:43   

         order_approved_at order_delivered_carrier_date  \
46     2018-03-15 10:29:33                          NaN   
288    2017-02-01 00:30:55                          NaN   
373    2018-05-13 17:11:58                          NaN   
383    2017-12-06 15:57:29                          NaN   
556    2017-07-12 15:04:10                          NaN   
...                    ...                          ...   
98267  2017-09-28 01:32:50                          NaN   
98319  2017-09-29 20:00:07                          NaN   
98551                  NaN                          NaN   
98642  2017-08-24 21:15:36                          NaN   
99105  2017-11-20 17:26:54                          NaN   

      order_delivered_customer_date order_estimated_delivery_date  
46                              NaN           2018-04-09 00:00:00  
288                             NaN           2017-03-10 00:00:00  
373                             NaN           2018-06-05 00:00:00  
383                             NaN           2018-01-03 00:00:00  
556                             NaN           2017-08-11 00:00:00  
...                             ...                           ...  
98267                           NaN           2017-10-27 00:00:00  
98319                           NaN           2017-10-17 00:00:00  
98551           

It doesn't seem to have an explanation for these missing values, we will merge dataframes keeping in mind that we will have some NaN values

In [137]:
data = pd.merge(data, order_items, on=['order_id'], how='outer')
data.head()

customer_id                customer_unique_id  \
0       06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1       18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2       4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3       b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4       4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                  ...                               ...   
113420  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
113421  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
113422  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
113423  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
113424  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

        customer_zip_code_prefix          customer_city customer_state  \
0                          14409                 franca             SP   
1                           9790  sao bernardo do campo             SP   
2                           1151              sao paulo             SP   
3                           8775        mogi das cruzes             SP   
4                          13056               campinas             SP   
...                          ...                    ...            ...   
113420                      3937              sao paulo             SP   
113421                      6764        taboao da serra             SP   
113422                     60115              fortaleza             CE   
113423                     92120                 canoas             RS   
113424                      6703                  cotia             SP   

                                order_id order_status  \
0       00e7ee1b050b8499577073aeb2a297a1    delivered   
1       29150127e6685892b6eab3eec79f59c7    delivered   
2       b2059ed67ce144a36e2aa97d2c9e9ad2    delivered   
3       951670f92359f4fe4a63112aa7306eba    delivered   
4       6b7d50bd145f6fc7f33cebabd7e49d0f    delivered   
...                                  ...          ...   
113420  6760e20addcf0121e9d58f2f1ff14298    delivered   
113421  9ec0c8947d973db4f4e8dcf1fbfa8f1b    delivered   
113422  fed4434add09a6f332ea398efd656a5c    delivered   
113423  e31ec91cea1ecf97797787471f98a8c2    delivered   
113424  28db69209a75e59f20ccbb5c36a20b90    delivered   

       order_purchase_timestamp    order_approved_at  \
0           2017-05-16 15:05:35  2017-05-16 15:22:12   
1           2018-01-12 20:48:24  2018-01-12 20:58:32   
2           2018-05-19 16:07:45  2018-05-20 16:19:10   
3           2018-03-13 16:06:38  2018-03-13 17:29:19   
4           2018-07-29 09:51:30  2018-07-29 10:10:09   
...                         ...                  ...   
113420      2018-04-07 15:48:17  2018-04-07 16:08:45   
113421      2018-04-04 08:20:22  2018-04-04 08:35:12   
113422      2018-04-08 20:11:50  2018-04-08 20:30:03   
113423      2017-11-03 21:08:33  2017-11-03 21:31:20   
113424      2017-12-19 14:27:23  2017-12-19 18:50:39   

       order_delivered_carrier_date order_delivered_customer_date  \
0               2017-05-23 10:47:57           2017-05-25 10:35:35   
1               2018-01-15 17:14:59           2018-01-29 12:41:19   
2               2018-06-11 14:31:00           2018-06-14 17:58:51   
3               2018-03-27 23:22:42           2018-03-28 16:04:25   
4               2018-07-30 15:16:00           2018-08-09 20:55:48   
...                             ...                           ...   
113420          2018-04-11 02:08:36           2018-04-13 20:06:37   
113421          2018-04-05 18:42:35           2018-04-11 18:54:45   
113422          2018-04-09 17:52:17           2018-05-09 19:03:15   
113423          2017-11-06 18:24:41           2017-11-16 19:58:39   
113424          2017-12-21 19:17:21           2017-12-26 18:42:36   

       order_est

In [138]:
data.isna().sum()

customer_id                         0
customer_unique_id                  0
customer_zip_code_prefix            0
customer_city                       0
customer_state                      0
order_id                            0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 161
order_delivered_carrier_date     1968
order_delivered_customer_date    3229
order_estimated_delivery_date       0
order_item_id                     775
product_id                        775
seller_id                         775
shipping_limit_date               775
price                             775
freight_value                     775
dtype: int64

## III - Data and order_reviews

In [140]:
order_reviews.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [139]:
order_reviews.shape

(99224, 7)

In [142]:
orders.shape

(99441, 8)

Should be the same number of rows

In [73]:
# Looking duplicates
order_reviews[order_reviews.duplicated(subset='order_id')]

review_id                          order_id  \
1119   46abf3ea0b2710ad41390fdb79c32d84  5040757d4e06a4be96d3827b860b4e7c   
3109   aa193e76d35950c4ae988237bb36ed2b  cf73e2cb1f4a9480ed70c154da3d954a   
8108   40294ea5a778dc62080d6b3f55d361ce  e1bc1083cd7acd30d0576335373b907d   
9064   32e2c7e889f7a185d462265398ee3631  c7cfea0c153e6382e32e84c2a9dd7d2e   
9795   95a3135743556b117d888cc8c6e12e11  f9c78e6e58306dc81efbbada1ac11f24   
...                                 ...                               ...   
98989  dfb3db02188d809d5cd199496b6da87e  c0db7d31ace61fc360a3eaa34dd3457c   
99108  2c6c08892b83ba4c1be33037c2842294  42ae1967f68c90bb325783ac55d761ce   
99164  2afe63a67dfd99b3038f568fb47ee761  c5334d330e36d2a810a7a13c72e135ee   
99178  44d1e9165ec54b1d89d33594856af859  a7dbcf5043158d6fa72859eead2f3d10   
99199  40743b46a0ee86375cedb95e82b78d75  3e93213bb8fdda91186b4018b2fe0030   

       review_score review_comment_title  \
1119              5                  NaN   
3109              5                  NaN   
8108              5                  NaN   
9064              5                  NaN   
9795              3                  NaN   
...             ...                  ...   
98989             5                  NaN   
99108             4                  NaN   
99164             5                  NaN   
99178             4                  NaN   
99199             5      OTIMA EMBALAGEM   

                                  review_comment_message review_creation_date  \
1119                                                 NaN  2017-11-07 00:00:00   
3109                                                 NaN  2018-01-18 00:00:00   
8108                                                 NaN  2018-03-23 00:00:00   
9064                                                 NaN  2017-10-21 00:00:00   
9795   O produto chegou antes do prazo mas recebi um ...  2017-09-14 00:00:00   
...                                                  ...                  ...   
98989                                                NaN  2018-02-17 00:00:00   
99108  Chegou um pouco amassada, mas nada de mais, e ...  2017-07-03 00:00:00   
99164  Muito bom, produto conforme anunciado, entrega...  2018-03-03 00:00:00   
99178                                                NaN  2017-05-24 00:00:00   
99199                                                NaN  2018-08-08 00:00:00   

      review_answer_timestamp  
1119      2017-11-10 20:07:48  
3109      2018-01-18 17:36:45  
8108      2018-03-24 00:23:06  
9064      2017-10-22 00:47:25  
9795      2017-09-17 11:14:08  
...                       ...  
98989     2018-02-19 19:29:19  
99108     2017-07-05 19:06:59  
99164     2018-03-04 22:56:47  
99178     2017-05-24 23:15:21  
99199     2018-08-08 16:56:16  

[551 rows x 7 columns]

In [81]:
order_reviews.loc[order_reviews['order_id'] == 'f9c78e6e58306dc81efbbada1ac11f24']

review_id                          order_id  \
4753  3c95105686e2ae71e8fd94b6483eba79  f9c78e6e58306dc81efbbada1ac11f24   
9795  95a3135743556b117d888cc8c6e12e11  f9c78e6e58306dc81efbbada1ac11f24   

      review_score review_comment_title  \
4753             2                  NaN   
9795             3                  NaN   

                                 review_comment_message review_creation_date  \
4753  comprei 4 cubas ao mesmo tempo, fui cobrada pe...  2017-09-19 00:00:00   
9795  O produto chegou antes do prazo mas recebi um ...  2017-09-14 00:00:00   

     review_answer_timestamp  
4753     2017-09-22 15:23:29  
9795     2017-09-17 11:14:08

In [60]:
# Looking duplicates
order_reviews[order_reviews.duplicated(subset='review_id')]

review_id                          order_id  \
3317   3242cc306a9218d0377831e175d62fbf  9c5bfba7de6a4abbb6ba0baab78d1622   
5719   308316408775d1600dad81bd3184556d  3fe4dbcdb046a475dbf25463c1ca78bd   
7213   8ee90ac383cf825bb7f4756130d4e74a  75d5d3d16567a27eefc5752aeb063072   
7805   d433c252647c51309432ca0b763f969b  191abde9c3bf77d1ce67d100d8db6d15   
9013   03a6a25db577d0689440933055111897  3fde8b7313af6b37b84b5c7594d7add0   
...                                 ...                               ...   
99108  2c6c08892b83ba4c1be33037c2842294  42ae1967f68c90bb325783ac55d761ce   
99124  6ec93e77f444e0b1703740a69122e35d  e1fdc6e9d1ca132377e862593a7c0bd4   
99164  2afe63a67dfd99b3038f568fb47ee761  c5334d330e36d2a810a7a13c72e135ee   
99167  017808d29fd1f942d97e50184dfb4c13  b1461c8882153b5fe68307c46a506e39   
99178  44d1e9165ec54b1d89d33594856af859  a7dbcf5043158d6fa72859eead2f3d10   

       review_score review_comment_title  \
3317              5                  NaN   
5719              5                  NaN   
7213              5                  NaN   
7805              1                  NaN   
9013              5                  NaN   
...             ...                  ...   
99108             4                  NaN   
99124             5                  NaN   
99164             5                  NaN   
99167             5                  NaN   
99178             4                  NaN   

                                  review_comment_message review_creation_date  \
3317                                              Ótimo   2017-07-21 00:00:00   
5719   Ajudem a rastrear uma das 3 prateleiras que fa...  2017-09-07 00:00:00   
7213                                           recomendo  2017-05-30 00:00:00   
7805                                                 NaN  2017-09-03 00:00:00   
9013   Muito Bom! Gostei Bastante! Tecido Ótimo! Aten...  2017-12-15 00:00:00   
...                                                  ...                  ...   
99108  Chegou um pouco amassada, mas nada de mais, e ...  2017-07-03 00:00:00   
99124              Vendedor compromisso do vou o cliente  2017-10-07 00:00:00   
99164  Muito bom, produto conforme anunciado, entrega...  2018-03-03 00:00:00   
99167                                                NaN  2018-03-02 00:00:00   
99178                                                NaN  2017-05-24 00:00:00   

      review_answer_timestamp  
3317      2017-07-22 01:06:37  
5719      2017-09-11 09:58:09  
7213      2017-06-02 18:16:17  
7805      2017-09-06 16:11:00  
9013      2017-12-16 01:32:18  
...                       ...  
99108     2017-07-05 19:06:59  
99124     2017-10-07 19:47:11  
99164     2018-03-04 22:56:47  
99167     2018-03-05 01:43:30  
99178     2017-05-24 23:15:21  

[814 rows x 7 columns]

In [79]:
order_reviews.loc[order_reviews['review_id'] == '3242cc306a9218d0377831e175d62fbf']

review_id                          order_id  \
1009  3242cc306a9218d0377831e175d62fbf  711246846695fa895c1776f0ddc30b20   
3317  3242cc306a9218d0377831e175d62fbf  9c5bfba7de6a4abbb6ba0baab78d1622   

      review_score review_comment_title review_comment_message  \
1009             5                  NaN                 Ótimo    
3317             5                  NaN                 Ótimo    

     review_creation_date review_answer_timestamp  
1009  2017-07-21 00:00:00     2017-07-22 01:06:37  
3317  2017-07-21 00:00:00     2017-07-22 01:06:37

We have sometimes :
- The same review_id for differents order_id
- The same order_id for differents review_id

Several reviews for the same order could be understandable if there are more than one product in the order. But the same review for different orders doesn't seem to make sense

In [145]:
order_reviews.loc[order_reviews['order_id'] == 'f9c78e6e58306dc81efbbada1ac11f24']

review_id                          order_id  \
4753  3c95105686e2ae71e8fd94b6483eba79  f9c78e6e58306dc81efbbada1ac11f24   
9795  95a3135743556b117d888cc8c6e12e11  f9c78e6e58306dc81efbbada1ac11f24   

      review_score review_comment_title  \
4753             2                  NaN   
9795             3                  NaN   

                                 review_comment_message review_creation_date  \
4753  comprei 4 cubas ao mesmo tempo, fui cobrada pe...  2017-09-19 00:00:00   
9795  O produto chegou antes do prazo mas recebi um ...  2017-09-14 00:00:00   

     review_answer_timestamp  
4753     2017-09-22 15:23:29  
9795     2017-09-17 11:14:08

In [146]:
order_items.loc[order_items['order_id'] == 'f9c78e6e58306dc81efbbada1ac11f24']

order_id  order_item_id  \
109914  f9c78e6e58306dc81efbbada1ac11f24              1   

                              product_id                         seller_id  \
109914  1d2d3b19f5c3d093d4dbccd3cd963a8f  7ddcbb64b5bc1ef36ca8c151f6ec77df   

        shipping_limit_date   price  freight_value  
109914  2017-09-07 22:23:53  194.99          20.85

We can see that one order with 2 reviews has only one products so it doesn't make sense as well : we will drop duplicates

In [149]:
order_reviews.drop_duplicates(subset=['order_id'], inplace=True)
order_reviews.drop_duplicates(subset=['review_id'], inplace=True)

In [150]:
data = pd.merge(data, order_reviews, on=['order_id'], how='outer')
data.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  \
0                     14409                 franca             SP   
1                      9790  sao bernardo do campo             SP   
2                      1151              sao paulo             SP   
3                      8775        mogi das cruzes             SP   
4                     13056               campinas             SP   

                           order_id order_status order_purchase_timestamp  \
0  00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1  29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2  b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3  951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4  6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   

     order_approved_at order_delivered_carrier_date  ...  \
0  2017-05-16 15:22:12          2017-05-23 10:47:57  ...   
1  2018-01-12 20:58:32          2018-01-15 17:14:59  ...   
2  2018-05-20 16:19:10          2018-06-11 14:31:00  ...   
3  2018-03-13 17:29:19          2018-03-27 23:22:42  ...   
4  2018-07-29 10:10:09          2018-07-30 15:16:00  ...   

                          seller_id  shipping_limit_date   price  \
0  7c67e1448b00f6e969d365cea6b010ab  2017-05-22 15:22:12  124.99   
1  b8bc237ba3788b23da09c0f1f3a3288c  2018-01-18 20:58:32  289.00   
2  7c67e1448b00f6e969d365cea6b010ab  2018-06-05 16:19:10  139.94   
3  7c67e1448b00f6e969d365cea6b010ab  2018-03-27 16:31:16  149.94   
4  4a3ca9315b744ce9f8e9374361493884  2018-07-31 10:10:09  230.00   

  freight_value                         review_id review_score  \
0         21.88  88b8b52d46df026a9d1ad2136a59b30b          4.0   
1         46.48  02fc48a9efa3e3d0f1a8ea26507eeec3          5.0   
2         17.79  5ad6695d76ee186dc473c42706984d87          5.0   
3         23.36  059a801bb31f6aab2266e672cab87bc5          5.0   
4         22.25  8490879d58d6c5d7773f2739a03f089a          5.0   

   review_comment_title                         review_comment_message  \
0                   NaN                                            NaN   
1                   NaN                                            NaN   
2                   NaN                                            NaN   
3                   NaN                                            NaN   
4         a melhor nota  O baratheon è esxelente Amo adoro o baratheon   

  review_creation_date  review_answer_timestamp  
0  2017-05-26 00:00:00      2017-05-30 22:34:40  
1  2018-01-30 00:00:00      2018-02-10 22:43:29  
2  2018-06-15 00:00:00      2018-06-15 12:10:59  
3  2018-03-29 00:00:00      2018-04-02 18:36:47  
4  2018-08-10 00:00:00      2018-08-17 01:59:52  

[5 rows x 24 columns]

## IV - Data and order_payments

In [151]:
order_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [153]:
order_payments.shape

(103886, 5)

In [154]:
# Looking duplicates
order_payments[order_payments.duplicated(subset='order_id')]

order_id  payment_sequential payment_type  \
1456    683bf306149bb869980b68d48a1bd6ab                   1  credit_card   
2324    e6a66a8350bb88497954d37688ab123e                   2      voucher   
2393    8e5148bee82a7e42c5f9ba76161dc51a                   1  credit_card   
2414    816ccd9d21435796e8ffa9802b2a782f                   1  credit_card   
2497    2cbcb371aee438c59b722a21d83597e0                   2      voucher   
...                                  ...                 ...          ...   
103778  fd86c80924b4be8fb7f58c4ecc680dae                   1  credit_card   
103817  6d4616de4341417e17978fe57aec1c46                   1  credit_card   
103860  31bc09fdbd701a7a4f9b55b5955b8687                   6      voucher   
103869  c9b01bef18eb84888f0fd071b8413b38                   1  credit_card   
103874  d41f1036712e18ee525939d08b7a1c98                   1  credit_card   

        payment_installments  payment_value  
1456                       1           8.58  
2324                       1          10.51  
2393                       1           0.67  
2414                       1           5.65  
2497                       1           7.80  
...                      ...            ...  
103778                     1          76.10  
103817                     1          19.18  
103860                     1          77.99  
103869                     6         238.16  
103874                     1           3.57  

[4446 rows x 5 columns]

In [167]:
# These duplicates could be explained by several payments for one order, let's check it

order_payments[order_payments.duplicated(subset=['order_id', 'payment_sequential'])]

Empty DataFrame
Columns: [order_id, payment_sequential, payment_type, payment_installments, payment_value]
Index: []

In [155]:
order_payments.shape[0] - order_payments[order_payments.duplicated(subset='order_id')].shape[0]

99440

In [157]:
orders.shape

(99441, 8)

In [160]:
# It seems to have only one row of difference, let's look at it

order_payments_unique = list(order_payments['order_id'].unique())
orders_unique = list(orders['order_id'].unique())

set(orders_unique) - set(order_payments_unique)

{'bfbd0f9bdef84302105ad712db648a6c'}

In [161]:
orders[orders['order_id'] == 'bfbd0f9bdef84302105ad712db648a6c']

order_id                       customer_id  \
30710  bfbd0f9bdef84302105ad712db648a6c  86dc2ffce2dfff336de2f386a786e574   

      order_status order_purchase_timestamp    order_approved_at  \
30710    delivered      2016-09-15 12:16:38  2016-09-15 12:16:38   

      order_delivered_carrier_date order_delivered_customer_date  \
30710          2016-11-07 17:11:53           2016-11-09 07:47:38   

      order_estimated_delivery_date  
30710           2016-10-04 00:00:00

This will create one NaN value during the merge

In [162]:
data = pd.merge(data, order_payments, on=['order_id'], how='outer')
data.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  \
0                     14409                 franca             SP   
1                      9790  sao bernardo do campo             SP   
2                      1151              sao paulo             SP   
3                      8775        mogi das cruzes             SP   
4                     13056               campinas             SP   

                           order_id order_status order_purchase_timestamp  \
0  00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1  29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2  b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3  951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4  6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   

     order_approved_at order_delivered_carrier_date  ...  \
0  2017-05-16 15:22:12          2017-05-23 10:47:57  ...   
1  2018-01-12 20:58:32          2018-01-15 17:14:59  ...   
2  2018-05-20 16:19:10          2018-06-11 14:31:00  ...   
3  2018-03-13 17:29:19          2018-03-27 23:22:42  ...   
4  2018-07-29 10:10:09          2018-07-30 15:16:00  ...   

                          review_id review_score  review_comment_title  \
0  88b8b52d46df026a9d1ad2136a59b30b          4.0                   NaN   
1  02fc48a9efa3e3d0f1a8ea26507eeec3          5.0                   NaN   
2  5ad6695d76ee186dc473c42706984d87          5.0                   NaN   
3  059a801bb31f6aab2266e672cab87bc5          5.0                   NaN   
4  8490879d58d6c5d7773f2739a03f089a          5.0         a melhor nota   

                          review_comment_message review_creation_date  \
0                                            NaN  2017-05-26 00:00:00   
1                                            NaN  2018-01-30 00:00:00   
2                                            NaN  2018-06-15 00:00:00   
3                                            NaN  2018-03-29 00:00:00   
4  O baratheon è esxelente Amo adoro o baratheon  2018-08-10 00:00:00   

  review_answer_timestamp  payment_sequential  payment_type  \
0     2017-05-30 22:34:40                 1.0   credit_card   
1     2018-02-10 22:43:29                 1.0   credit_card   
2     2018-06-15 12:10:59                 1.0   credit_card   
3     2018-04-02 18:36:47                 1.0   credit_card   
4     2018-08-17 01:59:52                 1.0   credit_card   

  payment_installments  payment_value  
0                  2.0         146.87  
1                  8.0         335.48  
2                  7.0         157.73  
3                  1.0         173.30  
4                  8.0         252.25  

[5 rows x 28 columns]